pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type CollectionId: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent the a time moment
+		type Moment: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// A way to obtain oracle values from feeders
+		type OracleProvider: ValueProvider<
+			(Self::AccountId, Self::CollectionId),
+			Self::OracleKey,
+			Value = Self::OracleValue,
+			Timestamp = Self::Moment,
+		>;
+
+		/// Used to verify collection admin permissions
+		type IsAdmin: PreConditions<(Self::AccountId, Self::CollectionId), Result = bool>;
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxFeedersPerKey: Get<u32>;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::OracleKey, (T::OracleValue, T::Moment), T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	/// Store the keys that are registed for this collection
+	/// Only keys registered in this stored can be used to create the collection
+	#[pallet::storage]
+	pub(crate) type Keys<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		Blake2_128Concat,
+		T::OracleKey,
+		KeyInfo<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AddedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		RemovedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		UpdatedFeeders {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		},
+		UpdatedCollection {
+			collection_id: T::CollectionId,
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// Collection size reached
+		MaxCollectionSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the feeders associated to an specific key.
+		///
+		/// NOTE: In the future this should be go though change guard.
+		/// By now we trust in the admin to choose honest feeders without
+		/// invertor supervision
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(2)]
+		pub fn update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
```

Comment:
Regarding the last requirement change in [this slack conversation]( https://kflabs.slack.com/archives/C050E3ZN2JU/p1701076174699669), do you think it still makes sense to let the admin to link the feeders to certain keys, or directly associate feeders to the whole collection?

Right now, because feeders are associated with the keys, the admin still handles which keys belong to the connection indirectly.

cc @hieronx @mustermeiszer

Reply:
Yes, let the admin decide and update here.